brainiak/utils/fmrisim.py

```diff
@@ -1382,7 +1478,7 @@ def generate_noise_volume(dimensions,
                                                              1)
     temporal_noise = temporal_noise - (temporal_noise_mean - spatial_noise)
```

Comment:
I think the combination of Line 1479 (where you essentially added spatial_noise to de-meaned temporal_noise) and 1482 (which adds spatial_noise again to temporal_noise) inflates the size of spatial_noise. I suggest changing Line 1479 to `temporal_noise = temporal_noise - (temporal_noise_mean)`. 
Further, if spatial_std means the std of noise for the mean intensity across voxels and temporal_std is the std of noise beyond the varation in spatial_noise, then demeaning temporal_noise already guarantees this. So I think you do not need the `if/else` clause in Line 1461 which attempts to correct the variance as you said.

Reply:
Yeah good catch, that is right, I am effectively doubling the spatial noise. 

However I am not sure I understand the second part. The variance of the spatial and temporal noise can be anything, if I don't put constraints on it I will get problems as described above

Reply:
Sorry then maybe I have misunderstanding of some definitions. Can you help me sort them out?

Function `noise_volume` generates an iid volume of noise with specified noise size and noise type, so both voxels in and out of brain have the same noise variance in `spatial_noise`. `temporal_noise` is also generated with `noise_volume` but then temporally demeaned. This means that its temporal average has no spatial variation, therefore contributes nothing to spatial noise. It is then added by `spatial_noise`, which essentiall biases the time series in each voxel by the value in the correspondent voxel in `spatial_noise`. Based on these, I assume `spatial_sd` means the size of variation in the mean intensity (averaged across time) across voxels, and `temporal_sd` means the temporal variation within each voxel relative to its mean intensity. I think your codes guarantee these, so it does not seem necessary to correct `spatial_noise` as in Line 1461. And I think `_calc_snr` seems also based on this definition. Lines like 2345 seem to be consistent with this definition.

But maybe you intended to use `spatial_sd` to mean `global_sd`, that is, the standard deviation of the whole 4-D matrix that includes the variance both in space and in time? In that case, I can see that you want to subtract the temporal variance from the variance of whole data to get a spatial_sd that follows the definition above. And with this definition what you did in Line 1461 makes sense. But then I think `_calc_snr` does not reflect this assumption because it estimates the real "spatial" standard deviation.

Reply:
Ah I think that is a clear explanation of what you are thinking. Let me clarify my thinking and then we can see if you agree. _generate_system_noise seeks to simulate the thermal noise properties of the scanner. This is added to the brain specific noise and also the baseline mean MR values of the brain (and skull and rest of the MR sensitive stuff in the FOV). The system noise has a big effect on the SNR since I am comparing the brain voxels to the non brain voxels. Critically, the SNR is calculated using only a single time point, not the average of the brain across time.

The system noise in this simulation (and real data) looks like the middle image below: 

![image](https://user-images.githubusercontent.com/11008543/43976163-bafa1696-9cad-11e8-89bb-7863fefa2014.png)

The idea is that each voxel has a mean activity with variability in time around that mean. The mean is mainly set by the baseline, but an additional component is the spatial noise that is added on top (parameterized by spatial_sd) in order to make sure there is variability between simulations (the baseline is identical between simulations with the same input/reference data). This spatial noise is consistent between time points. Added to the mean of each voxel is the temporal variability The temporal variability is characterized by temporal_sd. The idea of temporal_sd is simply to capture how thermal noise fluctuates across time. However, this temporal_sd is not there to affect/control SFNR but instead for SNR...

That may be confusing but let me explain. Remember we want SNR to reflect the brain mean to nonbrain variance at a given time point. Hence both the spatial_sd AND temporal_sd will contribute to the sd at any given timepoint (this wouldn't be true if SNR used the mean across time). Hence I want to ensure that there is a specific amount of variability for any given time point and to do that I need to constrain one of these two values not to exceed that target variability. I chose to constrain/limit spatial_sd but I could have done the opposite, I just think that makes less sense.

SFNR is less affected by all of this: the size of the variability in time accounted for by the system noise is constrained by 'temporal_proportion'.

Upon explaining this, I am questioning the necessity of spatial_sd, specifically whether the complexity it adds is justifiable. According to the logic, I could have only temporal_sd that is added to the baseline and that should achieve an appropriate simulation of the system noise. Although I think spatial_sd is potentially important theoretically, it may just make things unnecessarily complicated. 

Interested to hear your thoughts

Reply:
I see you point. So `spatial_sd` only matters when there are multiple runs. If there is a single run, it is undetectable since no one knows the real intensity at each voxel. Then I agree that temporal_noise is sufficient. It would only be a bit odd if the baseline outside of the brain are all set zero and temporal_noise is also demeaned - which makes every voxel's temporal mean strictly zero - which is unlikely but probably no one cares much. If you do not demean temporal_noise, then it seems fine.
About `_calc_snr`, I think it is called by `_fit_spatial` in 2362 with the input `noise`, and the `noise` fed to `_fit_spatial` in line 2761 seems to be a spatial-temporal matrix. So I think `_calc_snr` actually deals with 4D data. And I *think* the resulting SNR is based on the spatial variation of temporally demeaned noise.

Reply:
My sincerest apologies but in the process of this discussion and thinking about the code, I have realized that spatial_sd was necessary. Firstly, I made a mistake in what I said above: SNR is by default calculated based on the temporal average of all data:

_calc_snr does take in a 4D volume, however, one of its inputs is 'tr' which specifies the tr to be used for calculating the brain mean and nonbrain standard deviation. By default it uses all TRs (contrary to wait I said previously sorry). 

I have added some documentation and changed the names to make this more clear

The reason this matters is that the temporal_sd now doesn't affect the SNR (which is good to control for) since it is zero mean. But now if spatial_sd was zero then the spatial variance would simply be based on the template. As it turns out, that variability is too low in real data, which has more non-brain variance than the template (which makes sense because the template is a temporal average and so partial denoised). This means I need to add some amount of noise to the nonbrain voxels. Spatial_sd allows me to do this and _fit_spatial is the main way this is done. 

However, this does not mean that I am doing this correctly. For instance, I could remove the step where I subtract temporal_sd from spatial_sd. I do this, as described above, to limit the total variance of the system noise but that could be wrong. 

Happy to discuss further, this is a very thorny issue that I am not sure about.

Reply:
Great, I will remove the subtraction step and run through my thorough tests to see if it still works